### PR TITLE
Include a conditional update script

### PR DESCRIPTION
**Pull request by vorner**
Sorry for taking so long. This is for #2108.

I'm not sure if I'm really good at writing READMEs, so if its the wrong style or something, feel free to adjust or point it out.

### CODE REVIEW

File: rust-update [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+renice -n 10 -p $$
+
+export CARGO_INCREMENTAL=0
+export RUSTFLAGS='-C target-cpu=native'
+
+try_out() {
+	export RUSTUP_HOME=$HOME/.rustup-attempt
+	test -d $RUSTUP_HOME || (rustup toolchain add nightly && rustup default nightly)
+	rustup update
```
Comment (llogiq) [inline rust-update]: Should we explicitly `rustup update nightly`, as that is what clippy uses, and other toolchains could perhaps sneak in if someone does a rustup update in parallel in the same shell (yes, that's quite contrived, but better safe than sorry).

File: rust-update [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+	cargo install --force clippy
+	unset RUSTUP_HOME
+	export RUSTUP_HOME
+}
+
+[ "$1" = '-f' ] || try_out
+
+rustup update
```
Comment (llogiq) [inline rust-update]: Same here

File: README.md [hunk 3]
```diff
@@ -180,6 +180,14 @@ transparently:
 #[cfg_attr(feature = "cargo-clippy", allow(needless_lifetimes))]
 ```
 
+## Updating rustc
+
+Sometimes, rustc moves forward without clippy catching up. Therefore, if you
+updated rustc, clippy wouldn't work and wouldn't compile for a while.
```
Comment (llogiq) [inline README.md]: Perhaps "Therefore updating rustc may leave clippy in a non-functional state until we fix the resulting breakage."?

**Comment by llogiq** [comment]
Sorry for taking so long; I was on the move for a while and had restricted access to either my notebook or the network...

File: README.md (hunk 3 above)
Comment (vorner) [reply README.md]: OK

File: rust-update (hunk 1 above)
Comment (vorner) [reply rust-update]: I don't see what you mean „in the same shell“, as the script runs in its own instance of sh, so it's impossible to run any „user“ commands in it.

However, I think I guess the general direction of where you lead with the question. `rustup update` actually updates all the toolchains (and components and targets) if you don't specify one, which is *probably* what the user wants anyway (it is what *I* want for sure), no matter which one is your default.

On the other hand, it is not guaranteed that user's default toolchain is nightly, which this script assumed. I didn't modify the `rustup update` call, but I added `+nightly` in the cargo command, to make sure clippy is compiled with that.

Does that address your concern, or did I misunderstood the problem?

**Comment by oli-obk** [comment]
ping @llogiq looks like there was an update to this PR

File: rust-update (hunk 1 above)
Comment (llogiq) [reply rust-update]: No, that's fine.

**Comment by llogiq** [comment]
Revisiting this, we should probably warn that the script wasn't hardened and should only be run in a trusted environment.

Otherwise this looks good to me.

**Comment by vorner** [comment]
Does this expect any further action from me?

**Comment by oli-obk** [comment]
> Does this expect any further action from me?

Nope ;)

**Pull request merged**